### PR TITLE
<complex>: Fix exp and polar on NaN/infinity input

### PR DESCRIPTION
**Pull request by statementreply**
Works towards #887.

- [x] Fix `exp`.
  Also fix the imaginary part of `exp(complex{re, -0.0})`, which should be negative zero. The current implementation returns positive zero.
- [x] Fix `polar`.
  Most of those input cases (except `polar(+Inf, finite)`) are UB [on paper](http://eel.is/c++draft/complex.value.ops#9). This PR implements:
  - `polar(-rho, theta)` = `-polar(rho, theta)`
  - `polar(rho, theta)` ≈ `exp(complex{log(rho), theta})`, when `signbit(rho)` is `false`

### CODE REVIEW

File: stl/src/xlexp.cpp [hunk 1]
```diff
@@ -18,11 +18,14 @@ static const long double invln2 = 1.4426950408889634073599246810018921L;
 
 _CRTIMP2_PURE short __CLRCALL_PURE_OR_CDECL _LExp(
     long double* px, long double y, short eoff) { // compute y * e^(*px), (*px) finite, |y| not huge
-    if (*px < -hugexp || y == 0.0L) { // certain underflow
-        *px = 0.0L;
+    if (y == 0.0L) { // zero
+        *px = y;
+        return 0;
+    } else if (*px < -hugexp) { // certain underflow
+        *px = 0.0L * y;
```
Comment (StephanTLavavej) [inline stl/src/xlexp.cpp:25]: Why is this different from `xexp.cpp` which said `*px = _Xfe_underflow(y);`?

File: stl/src/xlexp.cpp [hunk 2]
```diff
@@ -18,11 +18,14 @@ static const long double invln2 = 1.4426950408889634073599246810018921L;
...
-        *px = 0.0L;
+    if (y == 0.0L) { // zero
+        *px = y;
+        return 0;
+    } else if (*px < -hugexp) { // certain underflow
+        *px = 0.0L * y;
         return 0;
     } else if (hugexp < *px) { // certain overflow
-        *px = _LInf._Long_double;
+        *px = _LInf._Long_double * y;
```
Comment (StephanTLavavej) [inline stl/src/xlexp.cpp:28]: Similarly, `xexp.cpp` said `*px = _Xfe_overflow(y);` - should this be consistent?

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]
Thanks for fixing this long-standing misbehavior! Please let me know if you'd like me to investigate how to make the code compatible with `/clr:pure` - I think that avoiding C++20 usage and dealing with the `float_control` warning should be sufficient, but I can verify that if you want.

File: stl/src/xmath.hpp [hunk 3]
```diff
@@ -165,4 +167,21 @@ long double* _LXp_sqrtx(long double*, int, long double*);
 
 _END_EXTERN_C_UNLESS_PURE
 
+// raise IEEE 754 exceptions
+#pragma float_control(except, on, push)
```
[the remark this answers is not in the document]
Comment (statementreply) [reply stl/src/xmath.hpp]: Guarded `#pragma float_control` by `#ifndef _M_CEE_PURE`.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I've pushed a merge with `main` - it was surprisingly free of manual merge conflicts, despite code being added near a removed shouty banner.

**Comment by StephanTLavavej** [comment]
I'm mirroring this to an MSVC-internal PR. It's totally fine to push changes for code review feedback, but please notify me in that case.

**Comment by StephanTLavavej** [comment]
Had to push one more change to fix the internal build - the `/clr:pure` build uses `/std:c++14` so we have to avoid terse `static_assert`.

**Review by StephanTLavavej — APPROVED** [review]

**Review by barcharcraz — APPROVED** [review]
This does appear to implement advertised behavior, although I mostly checked only that it did not do anything non-conforming (mess with the polar(0,x) case). I observe that the changes to polar are not currently covered by tests, since even after this change the polar llvm tests can not be enabled. We need to make sure to remember to turn those on.

**Comment by StephanTLavavej** [comment]
Thanks for fixing more of these complex bugs! :joy_cat: :lady_beetle: :tada:

**Pull request merged**